`test_keeper_reconfig_remove_many/test.py::test_reconfig_remove_2_and_leader` is flaky

[Test case history](https://play.clickhouse.com/play?user=play#U0VMRUNUCiAgICBjaGVja19zdGFydF90aW1lLCBjaGVja19uYW1lLCB0ZXN0X25hbWUsIHRlc3RfY29udGV4dF9yYXcsIHJlcG9ydF91cmwKRlJPTQogICAgY2hlY2tzCldIRVJFIDEKICAgIC0tIE9ubHkgaW4gbWFzdGVyCiAgICBBTkQgaGVhZF9yZWYgPSAnbWFzdGVyJwogICAgLS0gT25seSB1bnN1Y2Nlc3NmdWwgcnVucwogICAgQU5EIGNoZWNrX3N0YXR1cyAhPSAnc3VjY2VzcycKICAgIEFORCB0ZXN0X3N0YXR1cyBOT1QgSU4gKCdPSycsICdTS0lQUEVEJywgJycsICdCUk9LRU4nLCAnTk9UX0ZBSUxFRCcpCiAgICBBTkQgdGVzdF9uYW1lIExJS0UgJyV0ZXN0X2tlZXBlcl9yZWNvbmZpZ19yZW1vdmVfbWFueSUnCiAgICBBTkQgY2hlY2tfc3RhcnRfdGltZSA+PSBub3coKSAtIElOVEVSVkFMIDEgTU9OVEgKT1JERVIgQlkKICAgIGNoZWNrX3N0YXJ0X3RpbWUgREVTQw==)

Example failure (from [here](https://s3.amazonaws.com/clickhouse-test-reports/json.html?REF=master&sha=295824a87434274ecea38c45ee636a0fcf5984ca&name_0=MasterCI&name_1=Integration%20tests%20%28amd_asan%2C%20db%20disk%2C%20old%20analyzer%2C%203%2F6%29&name_1=Integration%20tests%20%28amd_asan%2C%20db%20disk%2C%20old%20analyzer%2C%203%2F6%29)):

```
File: test_keeper_reconfig_remove_many/test.py:119 - in test_reconfig_remove_2_and_leader
    ku.wait_configs_equal(config, zk2)
File: helpers/keeper_utils.py:436 - in wait_configs_equal
    right_config = sorted(get_config_str(right_zk).split("\n"))
File: helpers/keeper_utils.py:425 - in get_config_str
    return zk.get("/keeper/config")
File: helpers/keeper_utils.py:186 - in get
    return self.execute_query(f"get '{path}'", timeout)
File: helpers/keeper_utils.py:168 - in execute_query
    raise KeeperException(self.proc.stderr.readline().strip().decode())
E   helpers.keeper_utils.KeeperException: Code: 999. Coordination::Exception: Received error in heartbeat response: Operation timeout. (KEEPER_EXCEPTION), Stack trace (when copying this message, always include the lines below):
```

--- a/tests/integration/test_keeper_reconfig_remove_many/configs/keeper1.xml
+++ b/tests/integration/test_keeper_reconfig_remove_many/configs/keeper1.xml
@@ -1,5 +1,6 @@
 <clickhouse>
     <keeper_server>
+        <use_cluster>false</use_cluster>
         <tcp_port>9181</tcp_port>
         <server_id>1</server_id>
         <log_storage_path>/var/lib/clickhouse/coordination/log</log_storage_path>

--- a/tests/integration/test_keeper_reconfig_remove_many/configs/keeper1.xml
+++ b/tests/integration/test_keeper_reconfig_remove_many/configs/keeper1.xml
@@ -18,30 +18,35 @@
                 <id>1</id>
                 <hostname>node1</hostname>
                 <port>9234</port>
+                <priority>2</priority>
             </server>
             <server>
                 <id>2</id>
                 <hostname>node2</hostname>
                 <port>9234</port>
                 <start_as_follower>true</start_as_follower>
+                <priority>1</priority>
             </server>
             <server>
                 <id>3</id>
                 <hostname>node3</hostname>
                 <port>9234</port>
                 <start_as_follower>true</start_as_follower>
+                <priority>1</priority>
             </server>
             <server>
                 <id>4</id>
                 <hostname>node4</hostname>
                 <port>9234</port>
                 <start_as_follower>true</start_as_follower>
+                <priority>1</priority>
             </server>
             <server>
                 <id>5</id>
                 <hostname>node5</hostname>
                 <port>9234</port>
                 <start_as_follower>true</start_as_follower>
+                <priority>1</priority>
             </server>
         </raft_configuration>
     </keeper_server>

--- a/tests/integration/test_keeper_reconfig_remove_many/configs/keeper2.xml
+++ b/tests/integration/test_keeper_reconfig_remove_many/configs/keeper2.xml
@@ -18,30 +18,35 @@
                 <id>1</id>
                 <hostname>node1</hostname>
                 <port>9234</port>
+                <priority>2</priority>
             </server>
             <server>
                 <id>2</id>
                 <hostname>node2</hostname>
                 <port>9234</port>
                 <start_as_follower>true</start_as_follower>
+                <priority>1</priority>
             </server>
             <server>
                 <id>3</id>
                 <hostname>node3</hostname>
                 <port>9234</port>
                 <start_as_follower>true</start_as_follower>
+                <priority>1</priority>
             </server>
             <server>
                 <id>4</id>
                 <hostname>node4</hostname>
                 <port>9234</port>
                 <start_as_follower>true</start_as_follower>
+                <priority>1</priority>
             </server>
             <server>
                 <id>5</id>
                 <hostname>node5</hostname>
                 <port>9234</port>
                 <start_as_follower>true</start_as_follower>
+                <priority>1</priority>
             </server>
         </raft_configuration>
     </keeper_server>

--- a/tests/integration/test_keeper_reconfig_remove_many/configs/keeper2.xml
+++ b/tests/integration/test_keeper_reconfig_remove_many/configs/keeper2.xml
@@ -1,5 +1,6 @@
 <clickhouse>
     <keeper_server>
+        <use_cluster>false</use_cluster>
         <tcp_port>9181</tcp_port>
         <server_id>2</server_id>
         <log_storage_path>/var/lib/clickhouse/coordination/log</log_storage_path>

--- a/tests/integration/test_keeper_reconfig_remove_many/configs/keeper3.xml
+++ b/tests/integration/test_keeper_reconfig_remove_many/configs/keeper3.xml
@@ -1,5 +1,6 @@
 <clickhouse>
     <keeper_server>
+        <use_cluster>false</use_cluster>
         <tcp_port>9181</tcp_port>
         <server_id>3</server_id>
         <log_storage_path>/var/lib/clickhouse/coordination/log</log_storage_path>

--- a/tests/integration/test_keeper_reconfig_remove_many/configs/keeper3.xml
+++ b/tests/integration/test_keeper_reconfig_remove_many/configs/keeper3.xml
@@ -18,30 +18,35 @@
                 <id>1</id>
                 <hostname>node1</hostname>
                 <port>9234</port>
+                <priority>2</priority>
             </server>
             <server>
                 <id>2</id>
                 <hostname>node2</hostname>
                 <port>9234</port>
                 <start_as_follower>true</start_as_follower>
+                <priority>1</priority>
             </server>
             <server>
                 <id>3</id>
                 <hostname>node3</hostname>
                 <port>9234</port>
                 <start_as_follower>true</start_as_follower>
+                <priority>1</priority>
             </server>
             <server>
                 <id>4</id>
                 <hostname>node4</hostname>
                 <port>9234</port>
                 <start_as_follower>true</start_as_follower>
+                <priority>1</priority>
             </server>
             <server>
                 <id>5</id>
                 <hostname>node5</hostname>
                 <port>9234</port>
                 <start_as_follower>true</start_as_follower>
+                <priority>1</priority>
             </server>
         </raft_configuration>
     </keeper_server>

--- a/tests/integration/test_keeper_reconfig_remove_many/configs/keeper4.xml
+++ b/tests/integration/test_keeper_reconfig_remove_many/configs/keeper4.xml
@@ -18,30 +18,35 @@
                 <id>1</id>
                 <hostname>node1</hostname>
                 <port>9234</port>
+                <priority>2</priority>
             </server>
             <server>
                 <id>2</id>
                 <hostname>node2</hostname>
                 <port>9234</port>
                 <start_as_follower>true</start_as_follower>
+                <priority>1</priority>
             </server>
             <server>
                 <id>3</id>
                 <hostname>node3</hostname>
                 <port>9234</port>
                 <start_as_follower>true</start_as_follower>
+                <priority>1</priority>
             </server>
             <server>
                 <id>4</id>
                 <hostname>node4</hostname>
                 <port>9234</port>
                 <start_as_follower>true</start_as_follower>
+                <priority>1</priority>
             </server>
             <server>
                 <id>5</id>
                 <hostname>node5</hostname>
                 <port>9234</port>
                 <start_as_follower>true</start_as_follower>
+                <priority>1</priority>
             </server>
         </raft_configuration>
     </keeper_server>

--- a/tests/integration/test_keeper_reconfig_remove_many/configs/keeper4.xml
+++ b/tests/integration/test_keeper_reconfig_remove_many/configs/keeper4.xml
@@ -1,5 +1,6 @@
 <clickhouse>
     <keeper_server>
+        <use_cluster>false</use_cluster>
         <tcp_port>9181</tcp_port>
         <server_id>4</server_id>
         <log_storage_path>/var/lib/clickhouse/coordination/log</log_storage_path>

--- a/tests/integration/test_keeper_reconfig_remove_many/configs/keeper5.xml
+++ b/tests/integration/test_keeper_reconfig_remove_many/configs/keeper5.xml
@@ -1,5 +1,6 @@
 <clickhouse>
     <keeper_server>
+        <use_cluster>false</use_cluster>
         <tcp_port>9181</tcp_port>
         <server_id>5</server_id>
         <log_storage_path>/var/lib/clickhouse/coordination/log</log_storage_path>

--- a/tests/integration/test_keeper_reconfig_remove_many/configs/keeper5.xml
+++ b/tests/integration/test_keeper_reconfig_remove_many/configs/keeper5.xml
@@ -18,30 +18,35 @@
                 <id>1</id>
                 <hostname>node1</hostname>
                 <port>9234</port>
+                <priority>2</priority>
             </server>
             <server>
                 <id>2</id>
                 <hostname>node2</hostname>
                 <port>9234</port>
                 <start_as_follower>true</start_as_follower>
+                <priority>1</priority>
             </server>
             <server>
                 <id>3</id>
                 <hostname>node3</hostname>
                 <port>9234</port>
                 <start_as_follower>true</start_as_follower>
+                <priority>1</priority>
             </server>
             <server>
                 <id>4</id>
                 <hostname>node4</hostname>
                 <port>9234</port>
                 <start_as_follower>true</start_as_follower>
+                <priority>1</priority>
             </server>
             <server>
                 <id>5</id>
                 <hostname>node5</hostname>
                 <port>9234</port>
                 <start_as_follower>true</start_as_follower>
+                <priority>1</priority>
             </server>
         </raft_configuration>
     </keeper_server>

--- a/tests/integration/test_keeper_reconfig_remove_many/test.py
+++ b/tests/integration/test_keeper_reconfig_remove_many/test.py
@@ -9,11 +9,15 @@ import time
 import helpers.keeper_utils as ku
 from helpers.cluster import ClickHouseCluster, ClickHouseInstance
 
+from multiprocessing.dummy import Pool
+
 cluster = ClickHouseCluster(__file__)
 CONFIG_DIR = os.path.join(os.path.dirname(os.path.realpath(__file__)), "configs")
 
 nodes = [
-    cluster.add_instance(f"node{i}", main_configs=[f"configs/keeper{i}.xml"])
+    cluster.add_instance(
+        f"node{i}", main_configs=[f"configs/keeper{i}.xml"], stay_alive=True
+    )
     for i in range(1, 6)
 ]
 node1, node2, node3, node4, node5 = nodes
@@ -42,11 +46,38 @@ def create_client(node: ClickHouseInstance):
     )
 
 
+def start_clickhouse(node):
+    node.start_clickhouse()
+
+
+def setup_nodes(cluster):
+    for node in nodes:
+        node.stop_clickhouse()
+
+    p = Pool(len(nodes))
+    waiters = []
+    for node in nodes:
+        node.exec_in_container(["rm", "-rf", "/var/lib/clickhouse/coordination/log"])
+        node.exec_in_container(
+            ["rm", "-rf", "/var/lib/clickhouse/coordination/snapshots"]
+        )
+        node.exec_in_container(["rm", "-rf", "/var/lib/clickhouse/coordination/state"])
+
+        waiters.append(p.apply_async(start_clickhouse, (node,)))
+
+    for waiter in waiters:
+        waiter.wait()
+
+    ku.wait_nodes(cluster, nodes)
+
+
 def test_reconfig_remove_2_and_leader(started_cluster):
     """
     Remove 2 followers from a cluster of 5. Remove leader from 3 nodes.
     """
     global zk1, zk2, zk3, zk4, zk5
+
+    setup_nodes(started_cluster)
 
     zk1 = create_client(node1)
     config = ku.get_config_str(zk1)
@@ -117,7 +148,10 @@ def test_reconfig_remove_2_and_leader(started_cluster):
     zk2.stop()
 
     for i in range(100):
-        if any("leader" in ku.send_4lw_cmd(cluster, node, f"mntr") for node in [node2, node3]):
+        if any(
+            "leader" in ku.send_4lw_cmd(cluster, node, f"mntr")
+            for node in [node2, node3]
+        ):
             break
         time.sleep(1)
     else:

--- a/tests/integration/test_keeper_reconfig_remove_many/test.py
+++ b/tests/integration/test_keeper_reconfig_remove_many/test.py
@@ -4,6 +4,7 @@ import os
 import typing as tp
 
 import pytest
+import time
 
 import helpers.keeper_utils as ku
 from helpers.cluster import ClickHouseCluster, ClickHouseInstance
@@ -114,6 +115,14 @@ def test_reconfig_remove_2_and_leader(started_cluster):
     assert "node5" not in config
 
     zk2.stop()
+
+    for i in range(100):
+        if any("leader" in ku.send_4lw_cmd(cluster, node, f"mntr") for node in [node2, node3]):
+            break
+        time.sleep(1)
+    else:
+        assert False, "New leader not found"
+
     zk2 = create_client(node2)
     zk2.sync("/test_leader_0")
     ku.wait_configs_equal(config, zk2)

--- a/tests/integration/test_keeper_reconfig_remove_many/test.py
+++ b/tests/integration/test_keeper_reconfig_remove_many/test.py
@@ -147,6 +147,8 @@ def test_reconfig_remove_2_and_leader(started_cluster):
 
     zk2.stop()
 
+    # we just removed leader so we need to make sure new leader is elected
+    # before continuing with the checks
     for i in range(100):
         if any(
             "leader" in ku.send_4lw_cmd(cluster, node, f"mntr")

